kornia/augmentation/utils.py

```diff
@@ -39,3 +40,15 @@ def _validate_input_shape(input: torch.Tensor, channel_index: int, number: int)
         bool
     """
     return input.shape[channel_index] == number
+
+
+def _adapted_uniform(shape: Union[Tuple, torch.Size], low, high, same_on_batch=False):
+    r""" The uniform function that accepts 'same_on_batch'.
+    If same_on_batch is True, all values generated will be exactly same given a batch_size (shape[0]).
+    By default, same_on_batch is set to False.
+    """
```

Comment:
not for this PR, butcan we somehow create a logic so that the user specifies the dimension where wants to generate randomness ? I think that will go aligned with 3D data augmentation and temporal/video stuff /cc @mibaumgartner @justusschock
was thinking in having in the high level interface a `dim` param for this.
Besides, how we make this deterministic here ? using seed or generator ? /cc @anguelos

Reply:
To mention that I have given random param generator APIs with a new parameter ```same_on_batch```. In this way, the user can apply the exact same augmentation parameters across the whole batch.

In my opinion, 3D volumetric augmentation will be tricky since the Kornia functions were more or less considered to be applied to 2D images (BCHW) only. We need to do more evaluation first. For temporal/video, I think we can build on top of this implementation. To have a wrapper to somehow convert the (BTCHW) to *(bCHW), by b here, means to have the exact transforms along b axis.

Reply:
I think @shijianjian is right. Introducing support for volumetric data isn't that simple because most ops in kornia were designed for the 2D case and do not actively support BNDWH :) I'm not sure if a simple wrapper will do the job (maybe it's possible for videos but not "real" volumetric data where it is also possible to rotate around the D axes). Parameter injection for creating randomness is definitely possible there are other frameworks which use such a principle.

Reply:
We have had a quick discussion on this before. We can use a wrapper to play with the axis. I know it is a bit nasty.

> For semantic segmentation, it is actually a special case of BTCHW, which has a fixed T=2. User can input in (B, 2, C, H, W) with `apply_same_transform_on_dim=1`.

But the main idea is to use ```same_on_batch``` param to do for loop here. In semantic segmentation, we need every paired images to be the same transformed. Hence, a simple way is to have a wrapper for the for-loop. Just to give an rough idea here.

> In that logic, the `nn.Module` implementation shall be a wrapper to handle the dims, then for each functional transformation function, we gonna have a `apply_same_on_batch` param.
> 
> ```python
> if dim is not None:
>     x_t = torch.transpose(x, 0, dim)
>     x_t_transformed = torch.stack([transform(_x_t, apply_same_on_batch=True) for _x_t in x_t])
>     x_transformed = torch.transpose(x_t_transformed, dim, dim)
> else:
>     x_t = torch.reshape(x, (-1, *x.size()[-3:]))
>     x_t_transformed = torch.stack([transform(_x_t, apply_same_on_batch=False) for _x_t in x_t])
>     x_transformed = torch.reshape(*size())
> ```
> 
> I am not sure the use case of B**C**HW, etc. Basically, intensity-based transforms required to have 3 color channels, for cases like B**C**HW, I do not think those intensity-based transforms will work.

Reply:
I see, nice idea :) This should do the trick for videos where the same augmentation is applied on all frames. I think, this shouldn't work for "real" volumetric data, e.g. it is not possible to model a rotation around the depth axis (maybe it is possible to compute some kind of projection of the rotation for slice but this does not seem very straight forward and easy todo for more complex transforms) or did I miss something?

Reply:
Correct. This is more like a workaround to process any input that can be converted to BCHW format. BCHWD is not in this scope yet.

Also, we only generate parameters using for-loop, not doing augmentation in a for-loop. The last code snippet is outdated. Sorry.
```python
if dim is not None:
    x_t = torch.transpose(x, 0, dim)
    x_t_param = merge_param([get_param(_x_t, same_on_batch=True) for _x_t in x_t])
else:
    x_t = torch.reshape(x, (-1, *x.size()[-3:]))
    x_t_param = merge_param([get_param(_x_t, same_on_batch=False) for _x_t in x_t])

input_shape = input.shape
input_transformed = transform(input.reshape(-1, C, H, W), x_t_param)
input_transformed.reshape(input_shape)
```